examples/min-platform/embedding/wasmtime-platform.h

```diff
@@ -0,0 +1,203 @@
+ *
+ * Returns 0 if `wasmtime_longjmp` was used to return to this function.
+ * Returns 1 if `wasmtime_longjmp` was not called an `callback` returned.
+ */
+extern int32_t wasmtime_setjmp(const uint8_t **jmp_buf,
+                               void (*callback)(uint8_t*, uint8_t*),
+                               uint8_t *payload,
+                               uint8_t *callee);
+
+/**
+ * Paired with `wasmtime_setjmp` this is used to jump back to the `setjmp`
+ * point.
+ *
+ * The argument here was originally passed to `wasmtime_setjmp` through its
+ * out-param.
+ *
+ * This function cannot return.
+ *
+ * This function may be invoked from the `wasmtime_trap_handler_t`
+ * configured by `wasmtime_init_traps`.
+ */
+extern void wasmtime_longjmp(const uint8_t *jmp_buf);
+
+/**
+ * Initializes trap-handling logic for this platform.
+ *
+ * Wasmtime's implementation of WebAssembly relies on the ability to catch
+ * signals/traps/etc. For example divide-by-zero may raise a machine
+ * exception. Out-of-bounds memory accesses may also raise a machine
+ * exception. This function is used to initialize trap handling.
+ *
+ * The `handler` provided is a function pointer to invoke whenever a trap
+ * is encountered. The `handler` is invoked whenever a trap is caught by
+ * the system.
+ */
+extern void wasmtime_init_traps(wasmtime_trap_handler_t handler);
+
+/**
+ * Attempts to create a new in-memory image of the `ptr`/`len` combo which
+ * can be mapped to virtual addresses in the future. The returned
+ * `wasmtime_memory_image` pointer can be `NULL` to indicate that an image
+ * cannot be created. The structure otherwise will later be deallocated
+ * with `wasmtime_memory_image_free` and `wasmtime_memory_image_map_at`
+ * will be used to map the image into new regions of the address space.
+ */
+extern struct wasmtime_memory_image *wasmtime_memory_image_new(const uint8_t *ptr, uintptr_t len);
+
+/**
+ * Maps the `image` provided to the virtual address at `addr` and `len`.
+ *
+ * This semantically should make it such that `addr` and `len` looks the
+ * same as the contents of what the memory image was first created with.
+ * The mappings of `addr` should be private and changes do not reflect back
+ * to `wasmtime_memory_image`.
+ *
+ * In effect this is to create a copy-on-write mapping at `addr`/`len`
+ * pointing back to the memory used by the image originally.
+ *
+ * Aborts the process on failure.
+ */
+extern void wasmtime_memory_image_map_at(struct wasmtime_memory_image *image,
+                                         uint8_t *addr,
+                                         uintptr_t len);
+
+/**
+ * Replaces the VM mappings at `addr` and `len` with zeros.
+ *
+ * Aborts the process on failure.
+ */
+extern void wasmtime_memory_image_remap_zeros(struct wasmtime_memory_image *image,
+                                              uint8_t *addr,
+                                              uintptr_t len);
```

Comment:
Why does this need access to the memory image? It only modifies the VM mapping, right? And doesn't need to read the original content of the memory image.

Reply:
I was mostly just reflecting the existing Wasmtime API, which has not been meticulously thought out, onto the C API. You're right though that this entire function I think is not necessary, and I've replaced it with a call to `wasmtime_mmap_remap`.